apps/docusaurus/docs/guides/randomness.md

```diff
@@ -0,0 +1,314 @@
+
+- private, and
+- annotated with `#[randomness]`.
+
+It's now a good time to think about what user actions need randomness API, write them down, and make sure they are private and have the right attribute, as shown in the example below.
+
+```rust
+module module_owner::lottery {
+    // ...
+
+    #[randomness]
+    entry fun decide_winner() {
+        // ...
+    }
+}
+```
+
+### Call the API
+
+The APIs are public functions under `0x1::randomness` and can be referenced directly, as demonstrated in the lottery example above.
+
+```rust
+module module_owner::lottery {
+    // ...
+
+    #[randomness]
+    entry fun decide_winner() {
+        // ...
+        let winner_idx = aptos_framework::randomness::u64_range(0, n);
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+The above example uses function `u64_range()` but many other basic types are also supported.
+Here's a quick overview of all the API, where `T` can be one of `u8, u16, u32, u64, u128, u256`.
+
+```
+/// Generates an T uniformly at random.
+fun T_integer(): T
+
+/// Generates a number `[min_incl, max_excl)` uniformly at random.
+fun T_range(min_incl, ): T
+
+/// Generates a sequence of bytes uniformly at random
+fun bytes(n: u64): vector<u8>;
+
+/// Generate a permutation of `[0, 1, ..., n-1]` uniformly at random.
+/// If n is 0, returns the empty vector.
+fun permutation(n: u64): vector<u64>;
+```
+
+### Security warning: randomness API calls in non-entry functions
+
+As your dApp gets more complicated, you may have multiple entry functions that need to share the same randomness-dependent logic, and want to pull the logic out as a separate helper function.
+
+While this is supported as shown below, extra care must be taken.
+
+```rust
+module module_owner::lottery {
+    // ...
+
+    #[randomness]
+    entry fun decide_winner_v0() {
+        // ...
+        decide_winner_internal(lottery_state);
+    }
+
+    #[randomness]
+    entry fun decide_winner_v1() {
+        // ...
+        decide_winner_internal(lottery_state);
+    }
+
+    // A private helper function
+    fun decide_winner_internal(lottery_state: &mut lottery_state) {
+        let n = std::vector::length(&lottery_state.players);
+        let winner_idx = aptos_framework::randomness::u64_range(0, n);
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+If `decide_winner_internal()` were accidentally marked public,
+malicious players can deploy their own contract to:
+
+1. call`decide_winner_internal()`;
+1. read the lottery result (assuming the `lottery` module has some getter functions for the result);
+1. abort if the result is not in their favor.
+   By repeatedly calling their own contract until a txn succeeds,
+   malicious users can bias the uniform distribution of the winner (dApp developer's initial design).
+   This is referred to as a _test-and-abort attack_.
```

Comment:
Add link to AIP-41: https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-41.md#test-and-abort-attacks

Reply:
fixed